### PR TITLE
*: hotfix panic from tokio-timer

### LINKED ISSUE #11940: TiKV running over 2 years may panic
### What version of TiKV are you using?
`nightly`(this case also appears on v3.0.5)

### Steps to reproduce

Running TiKV for 795 days, then some RPC failure may cause panic.

A simple reproduce using [Chaos Mesh](https://github.com/chaos-mesh/chaos-mesh/):

```bash
tiup playground --tiflash=0 --monitor=false
sudo kill {pid_of_pd}
sudo ./watchmaker -pid {pid_of_tikv} -sec_delta 68719436 -clk_ids=CLOCK_MONOTONIC
```

### What did you expect?

No fatal error with TiKV.

### What did happened?

TiKV quits with fatal error.

```log
[2022/02/08 20:32:34.418 +08:00] [ERROR] [util.rs:419] ["request failed, retry"] [err_code=KV:PD:Unknown] [err="Other(\"[components/pd_client/src/tso.rs:85]: TimestampRequest channel is closed\")"]
[2022/02/08 20:32:34.418 +08:00] [ERROR] [util.rs:419] ["request failed, retry"] [err_code=KV:PD:Unknown] [err="Other(\"[components/pd_client/src/tso.rs:85]: TimestampRequest channel is closed\")"]
[2022/02/08 20:32:34.418 +08:00] [ERROR] [util.rs:419] ["request failed, retry"] [err_code=KV:PD:Unknown] [err="Other(\"[components/pd_client/src/tso.rs:85]: TimestampRequest channel is closed\")"]
[2022/02/08 20:32:34.418 +08:00] [ERROR] [util.rs:419] ["request failed, retry"] [err_code=KV:PD:Unknown] [err="Other(\"[components/pd_client/src/tso.rs:85]: TimestampRequest channel is closed\")"]
[2022/02/08 20:32:34.421 +08:00] [FATAL] [lib.rs:465] ["index out of bounds: the len is 6 but the index is 6"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:464:18\n   1: std::panicking::rust_panic_with_hook\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:626:17\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:519:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/sys_common/backtrace.rs:141:18\n   4: rust_begin_unwind\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:515:5\n   5: core::panicking::panic_fmt\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/panicking.rs:92:14\n   6: core::panicking::panic_bounds_check\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/panicking.rs:69:5\n   7: <usize as core::slice::index::SliceIndex<[T]>>::index_mut\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/slice/index.rs:190:14\n      core::slice::index::<impl core::ops::index::IndexMut<I> for [T]>::index_mut\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/slice/index.rs:26:9\n      <alloc::vec::Vec<T,A> as core::ops::index::IndexMut<I>>::index_mut\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/alloc/src/vec/mod.rs:2445:9\n      tokio_timer::wheel::Wheel<T>::insert\n             at /rust/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.13/src/wheel/mod.rs:114:9\n      tokio_timer::timer::Timer<T,N>::add_entry\n             at /rust/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.13/src/timer/mod.rs:324:15\n   8: tokio_timer::timer::Timer<T,N>::process_queue\n             at /rust/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.13/src/timer/mod.rs:301:21\n   9: <tokio_timer::timer::Timer<T,N> as tokio_executor::park::Park>::park\n             at /rust/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.13/src/timer/mod.rs:361:9\n      tokio_timer::timer::Timer<T,N>::turn\n             at /rust/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.13/src/timer/mod.rs:256:21\n  10: tikv_util::timer::start_global_timer::{{closure}}\n             at /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tikv/components/tikv_util/src/timer.rs:98:17\n  11: std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/sys_common/backtrace.rs:125:18\n  12: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/thread/mod.rs:476:17\n  13: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panic.rs:347:9\n  14: std::panicking::try::do_call\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:401:40\n      std::panicking::try\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:365:19\n      std::panic::catch_unwind\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panic.rs:434:14\n      std::thread::Builder::spawn_unchecked::{{closure}}\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/thread/mod.rs:475:30\n      core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/ops/function.rs:227:5\n  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/alloc/src/boxed.rs:1572:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/alloc/src/boxed.rs:1572:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/sys/unix/thread.rs:91:17\n  16: start_thread\n  17: __GI___clone\n"] [location=/rust/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.13/src/wheel/mod.rs:114] [thread_name=timer]
```

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### What is changed and how it works?

Issue Number: Close #11940

What's Changed:

```commit-message
*: hotfix panic from tokio-timer by enlarging the length of the level vector
```

`tokio-timer 0.2.13`'s time wheel works when the elapsed time since the process starting is less than `2^36 ms`(about 795 days), otherwise, it may panic. This PR switch `tokio-timer` to a patch version which enlarges the length of the level vector. In this patch, it supports the elapsed time less than `2^42 ms`(about 139 years), which should be long enough.

The changes can be found here: https://github.com/you06/tokio-timer-0.2.13/compare/0c5fb54...35090cc

### Related changes

- Need to cherry-pick to the release branch
  - release-4.0
  - release-5.0
  - release-5.1
  - release-5.2
  - release-5.3
  - release-5.4

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

The test steps are in #11940 which is based on chaos mesh. I did the following test for this PR.

```sh
sudo ./watchmaker -pid {pid_of_tikv} -sec_delta 68719436 -clk_ids=CLOCK_MONOTONIC # simulate time drift +795 days, tikv works fine
sudo ./watchmaker -pid {pid_of_tikv} -sec_delta 4366510511 -clk_ids=CLOCK_MONOTONIC # simulate time drift +138 years, tikv works fine
sudo ./watchmaker -pid {pid_of_tikv} -sec_delta 4398046411 -clk_ids=CLOCK_MONOTONIC # simulate time drift +139 years, tikv panics
```

Side effects

### Release note 

```release-note
Fixes the bug that TiKV keep running over 2 years may panic.
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -149,7 +149,7 @@ coprocessor_plugin_api = { path = "components/coprocessor_plugin_api" }
 time = "0.1"
 tipb = { git = "https://github.com/pingcap/tipb.git" }
 tokio = { version = "1.12", features = ["full"] }
-tokio-timer = "0.2"
+tokio-timer = { git = "https://github.com/you06/tokio-timer-0.2.13.git", branch = "master" }
```
Comment (BusyJay) [inline Cargo.toml:152]: Should be forked to tikv org and changes need to get two LGTMs.

Comment (you06) [reply Cargo.toml:152]: This dependency needs to be removed once the self-maintained time library that you mentioned in the issue is ready, so it's a temporary solution only. Do we need to fork this crate into tikv org?

Comment (BusyJay) [reply Cargo.toml:152]: Yes.

Comment (you06) [reply Cargo.toml:152]: I don't have the permission of creating a repo in TiKV org. Can you help me create a repo and forward my commits into it?

Comment (BusyJay) [reply Cargo.toml:152]: Now it's at https://github.com/tikv/tokio/tree/tokio-timer-hotfix.

**Review by youjiali1995 — APPROVED** [review]

**Review by BusyJay — APPROVED** [review]

**Pull request merged**